### PR TITLE
Support for single-quote bracket notation for params

### PR DESCRIPTION
**Pull request by skaegi**
This change adds single-quote bracket notation to the work done in #4215. This is consistent with how referencing is done elsewhere in K8s in the downwards api.

The original patch used the name subscript notation to describe this however the standard name for this approach is bracket notation and updates the doc accordingly.

This is based on TEP: tektoncd/community#503 -- which also needs to be aligned with the two bracket notation approaches in the implementation. See https://github.com/tektoncd/community/pull/527

# Changes

This change adds single-quote bracket param matching to the params and param results analagous with the double-quote bracket param matching. It also adds similar test cases and updates the doc with the single-quote syntax. 

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in or deleted (only if no user facing changes)

# Release Notes

```release-note
Params and result can now be referenced with bracket notation using both single and double quotes in addition to dot notation. For example, the following are equivalent:  $(param.myparam), $(param['myparam']), and $(param["myparam"]). Bracket notation has the additional benefit of allowing users to work with parameter names containing conflicting characters like "." (e.g. $(param['my.param']) or $(param["my.param"]).
```

### CODE REVIEW

File: pkg/reconciler/taskrun/resources/apply_test.go
Comment (pritidesai) [inline pkg/reconciler/taskrun/resources/apply_test.go:1166]: Please add an additional step to include the bracket notation 🙏

**Comment by pritidesai** [comment]
thanks @skaegi for implementing this feature, I have left a comment to add one unit test and will be great if you can update the [proposal](https://github.com/tektoncd/community/blob/main/teps/0080-support-domainscoped-parameterresult-names.md#proposal) with the bracket notation:

Proposal has:

```yaml
  steps:
    - image: $(params[dev.mattmoor.foo])
```

Correct it to use bracket notation with double quotes:

```yaml
  steps:
    - image: $(params["dev.mattmoor.foo"])
```

And bracket notation with single quotes:

```yaml
  steps:
    - image: $(params['dev.mattmoor.foo'])
```

/approve

**Comment by pritidesai** [comment]
Putting on a brief hold to collect feedback from the other maintainers, have added to the [agenda](https://docs.google.com/document/d/17PodAxG8hV351fBhSu7Y_OIPhGTVgj6OJ2lPphYYRpU/edit#heading=h.wq43sf1qrb1i) in the next API WG.

thanks @skaegi for your patience 🙏 

/hold

**Comment by pritidesai** [comment]
thanks @skaegi for updating the [proposal](https://github.com/tektoncd/community/pull/527) with the correct syntax. Based on the discussion in the API WG (10/04), I am cancelling the hold 🙏 

/hold cancel

**Review by vdemeester — APPROVED** [review]

**Comment by pritidesai** [comment]
thanks @vdemeester for approving, changing my approval to lgtm 🙏 

/lgtm

**Comment by jerop** [comment]
thanks @skaegi 

/lgtm

**Pull request merged**
